### ISSUE TITLE
Keypoints to 2dBoundingBox/Segmentation

### ISSUE THREAD

**Issue by GiorgioPasini** [Unity-Technologies/com.unity.perception#349]
Hello there, 
thank you really much for this awesome package.

For a project I need to obtain two bounding boxes, (maybe even segmentation if possible), one for the body and the other for the head of the character.
Given the knowledge acquired in the tutorial a first approach might be to actually have two distinct objects and just go on with that. The thing is that I think it would be quite a messy solution, prone to all kinds of issues and inefficiencies.
I was therefore thinking that a better way might be a hybrid solution between the already-implemented boundingBox/segmentation labellinga and the KeypointLabeler (i.e. if I can get where the head is through keypoints then I should be able to extract a bounding box as well).
So the question is: how do I achieve this? Maybe by accessing the weightpainting of the rig to isolate the vertices of the head and the body and from there getting the boundingBox/segmentation labelling? I think the already-implemented labelers work something like that but I have no other clue.

Thanks so much for the help and again thanks for the awesome work done so far.

**Comment by GiorgioPasini** [comment]
Hello there, 
thank you really much for this awesome package.

For a project I need to obtain two bounding boxes, (maybe even segmentation if possible), one for the body and the other for the head of the character.
Given the knowledge acquired in the tutorial a first approach might be to actually have two distinct objects and just go on with that. The thing is that I think it would be quite a messy solution, prone to all kinds of issues and inefficiencies.
I was therefore thinking that a better way might be a hybrid solution between the already-implemented boundingBox/segmentation labelling and the KeypointLabeler (i.e. if I can get where the head is through keypoints then I should be able to extract a bounding box as well).
So the question is: how do I achieve this? Maybe by accessing the weightpainting of the rig to isolate the vertices of the head and the body and from there getting the boundingBox/segmentation labelling? I think the already-implemented labelers work something like that but I have no other clue.

Thanks so much for the help and again thanks for the awesome work done so far.

**Comment by RuiyuZ** [comment]
Hi @GiorgioPasini, I think your hybrid solution is going to be quite hard to implement with the current perception package because we can't separate the skinned mesh of the human model into different labeled parts. If you want to give it a try you can modify the segmentation shader or copy-paste the labeler to do the texture approach (which is similar to their "weight painting the rig" idea. But it is gonna be not simple. 

For your purpose though, If you just want to separate the head and the body. I think you can start by using invisible extra objects for the head, and the rest of the body could just use the regular segmentation for the geometry.

![head](https://user-images.githubusercontent.com/37159720/128570686-58830c43-4dd2-4814-b51c-c5efae3152ca.gif)

**Comment by GiorgioPasini** [comment]
Oh I see, thank you for the answer!
So you would suggest to start by "tricking" the labeler into labeling the head by making it label an invisible object placed in correspondence of the head instead? That seems reasonable and I think it would be a good compromise. 

How do you suggest to make it invisible to not let it interfere with the actual final image needed for the training? 

I would think of applying a totally transparent shader but I have the impression there might be an easier/more straightforward option somewhere. 

Thank you again for the help!

**Comment by RuiyuZ** [comment]
Hi @GiorgioPasini, to make an object invisible you can turn off the rendering of a GameObject by disabling its MeshRenderer component, e.g

`GetComponent(MeshRenderer).enabled = false;`

This way you can make it invisible but still interact with it.

Hope this helps :)

**Comment by GiorgioPasini** [comment]
![labellingAndTransparency_gif](https://user-images.githubusercontent.com/64695113/134196824-d24a69b5-ac47-4a6b-b8fe-aaff2bc2b869.gif)

I tried to disable the MeshRenderer but it appears that by doing that even the labeler stops working. Fortunately making a transparent material was way way easier than expected, so by simply using a transparent one the problem has been solved. Another curious thing that I've noticed is that the existence of another labelled object seems to interact with the labelling of the one "behind it" even if there's no actual visual impediment.
At first it seemed that the portion "inside" the transparent object would be cleanly excluded by the labelling of the rest of the object (the other label you see in the GIF is indeed a label applied to the whole model head included), but after rotating the camera it seemed that said exclusion wasn't that clean, so maybe something more is happening under the hood.
Posted it so maybe someone checking the thread could maybe benefit from it.

So this actually seem to be a first solution to the 2D Box labelling of separate parts of the same model. I guess animation-wise, instead of having the capsule to be the transparent target for the box labelling, an actual lowpoly version of the same mesh, bound to the same rig in the same way but simply with transparent material, would be even better. 

I'll try to see what happens with the segmentation part of the issue
